canvas/image.go

```diff
@@ -51,38 +59,171 @@ func (i *Image) Alpha() float64 {
 // and height parameters passed to pixelColor.
 func NewRaster(pixelColor func(x, y, w, h int) color.Color) *Image {
 	return &Image{
-		PixelColor: pixelColor,
+		Raster: func(w, h int) image.Image {
+			// raster first pixel, figure out color type
+			var dst draw.Image
+			rect := image.Rect(0, 0, w, h)
+			switch pixelColor(0, 0, w, h).(type) {
+			case color.Alpha:
+				dst = image.NewAlpha(rect)
+			case color.Alpha16:
+				dst = image.NewAlpha16(rect)
+			case color.CMYK:
+				dst = image.NewCMYK(rect)
+			case color.Gray:
+				dst = image.NewGray(rect)
+			case color.Gray16:
+				dst = image.NewGray16(rect)
+			case color.NRGBA:
+				dst = image.NewNRGBA(rect)
+			case color.NRGBA64:
+				dst = image.NewNRGBA64(rect)
+			case color.RGBA:
+				dst = image.NewRGBA(rect)
+			case color.RGBA64:
+				dst = image.NewRGBA64(rect)
+			default:
+				dst = image.NewRGBA(rect)
+			}
+
+			for x := 0; x < w; x++ {
+				for y := 0; y < h; y++ {
+					dst.Set(x, y, pixelColor(x, y, w, h))
+				}
+			}
+
+			return dst
+		},
 	}
 }
 
+type subImg interface {
+	SubImage(r image.Rectangle) image.Image
+}
+
 // NewRasterFromImage returns a new Image instance that is rendered from the Go
 // image.Image passed in.
 // Images returned from this method will map pixel for pixel to the screen
 // starting img.Bounds().Min pixels from the top left of the canvas object.
+// Truncates rather than scales the image.
```

Comment:
I think we should expand on this - what, for example, happens if the icon is too small?
Perhaps "If the image is too large it will be truncated, if it is too small then transparent pixels will be added to the bottom and right edges."?

Reply:
So, at the moment I realized that transparent pixels will not always be added if the user provides an image format not supporting transparency. If the image format does not support transparency, it will add zero-pixels (usually black).

Reply:
Good point, seems like behaviour we could document - and even test for :)

Reply:
So do you think that I should be converting to transparency-supporting formats if I have to do this?

Reply:
Actually, scaling & padding is no longer within the scope of this function.

Reply:
I think your idea of supporting the image format where possible is a great idea - we just need to make sure the documentation matches wherever this lands :)